pkg/controllers/vdb/promotesandboxtomain_reconciler.go

```diff
@@ -0,0 +1,174 @@
+	initiator, ok := s.PFacts.findFirstPodSorted(func(v *PodFact) bool {
+		return v.isPrimary && v.upNode
+	})
+	if !ok {
+		s.Log.Info("No Up nodes found. Requeue reconciliation.")
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	s.VRec.Eventf(s.Vdb, corev1.EventTypeNormal, events.PromoteSandboxToMainStart,
+		"Starting promote sandbox %q to main", sandboxName)
+
+	err := s.Dispatcher.PromoteSandboxToMain(ctx,
+		promotesandboxtomain.WithInitiator(initiator.podIP),
+		promotesandboxtomain.WithSandbox(sandboxName),
+	)
+	if err != nil {
+		s.VRec.Eventf(s.Vdb, corev1.EventTypeWarning, events.PromoteSandboxToMainFailed,
+			"Failed to promote sandbox %q to main", sandboxName)
+		return ctrl.Result{}, err
+	}
+	s.VRec.Eventf(s.Vdb, corev1.EventTypeNormal, events.PromoteSandboxToSucceeded,
+		"Successfully promote sandbox %q to main", sandboxName)
+
+	return ctrl.Result{}, nil
+}
+
+// updateSandboxScTypeInVdb update SandboxPrimarySubcluster to PrimarySubcluster in vdb.spec.subclusters
+// and remove sandbox spec and status in vdb after promoting to main
+func (s *PromoteSandboxToMainReconciler) updateSandboxScTypeInVdb(ctx context.Context, sandboxName string) error {
+	scSbMap := s.Vdb.GenSubclusterSandboxMap()
+
+	// remove sandbox in spec
+	for i := len(s.Vdb.Spec.Sandboxes) - 1; i >= 0; i-- {
+		_, err := vk8s.UpdateVDBWithRetry(ctx, s.VRec, s.Vdb, func() (bool, error) {
+			if s.Vdb.Spec.Sandboxes[i].Name == sandboxName {
+				s.Vdb.Spec.Sandboxes[i].Subclusters = []vapi.SubclusterName{}
+				s.Vdb.Spec.Sandboxes = append(s.Vdb.Spec.Sandboxes[:i], s.Vdb.Spec.Sandboxes[i+1:]...)
+			}
+			return true, nil
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// update sandboxPrimarySubcluster to primarySubcluster in spec
+	for sc, sb := range scSbMap {
+		if sb == sandboxName {
+			_, err := vk8s.UpdateVDBWithRetry(ctx, s.VRec, s.Vdb, func() (bool, error) {
+				for j := range s.Vdb.Spec.Subclusters {
+					if s.Vdb.Spec.Subclusters[j].Name == sc && s.Vdb.Spec.Subclusters[j].Type == vapi.SandboxPrimarySubcluster {
+						s.Vdb.Spec.Subclusters[j].Type = vapi.PrimarySubcluster
+					}
+				}
+				return true, nil
+			})
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// remove sandbox in status
+	unsandboxSbScMap := s.Vdb.GenSandboxSubclusterMapForUnsandbox()
+	unsandboxedScNames, found := unsandboxSbScMap[sandboxName]
+	if !found {
+		s.Log.Info("the sandbox inside it does not need to be removed")
+		return nil
+	}
+	updateStatus := func(vdbChg *vapi.VerticaDB) error {
+		// update the sandbox's subclusters in sandbox status
+		for i := len(vdbChg.Status.Sandboxes) - 1; i >= 0; i-- {
+			if vdbChg.Status.Sandboxes[i].Name != sandboxName {
+				continue
+			}
+			vdbChg.Status.Sandboxes[i].Subclusters = vutil.SliceDiff(vdbChg.Status.Sandboxes[i].Subclusters, unsandboxedScNames)
+			if len(vdbChg.Status.Sandboxes[i].Subclusters) == 0 {
```

Comment:
You don't need this. We know that promote sandbox will "unsandbox" all the subclusters so we should just remove the sandbox from status like we do with spec.